external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutAbstractTest.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.storm.kafka.spout.internal.KafkaConsumerFactoryDefault;
+import org.apache.storm.spout.SpoutOutputCollector;
+import org.apache.storm.task.TopologyContext;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.utils.Time;
+import org.junit.Before;
+import org.junit.Rule;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.MockitoAnnotations;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public abstract class KafkaSpoutAbstractTest {
+    @Rule
+    public KafkaUnitRule kafkaUnitRule = new KafkaUnitRule();
+
+    @Captor
+    ArgumentCaptor<Map<TopicPartition, OffsetAndMetadata>> commitCapture;
+
+    final TopologyContext topologyContext = mock(TopologyContext.class);
+    final Map<String, Object> conf = new HashMap<>();
+    final SpoutOutputCollector collector = mock(SpoutOutputCollector.class);
+    final long commitOffsetPeriodMs = 2_000;
+    final int maxRetries = 3;
+    KafkaConsumer<String, String> consumerSpy;
+    KafkaConsumerFactory<String, String> consumerFactory;
+    KafkaSpout<String, String> spout;
+    final int maxPollRecords = 10;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+
+        final KafkaSpoutConfig<String, String> spoutConfig = createSpoutConfig();
+
+        consumerSpy = spy(new KafkaConsumerFactoryDefault<String, String>().createConsumer(spoutConfig));
+
+        consumerFactory = new KafkaConsumerFactory<String, String>() {
+            @Override
+            public KafkaConsumer<String, String> createConsumer(KafkaSpoutConfig<String, String> kafkaSpoutConfig) {
+                return consumerSpy;
+            }
+
+        };
+
+        spout = new KafkaSpout<>(spoutConfig, consumerFactory);
+    }
+
+
+    abstract KafkaSpoutConfig<String, String> createSpoutConfig();
+
+    void prepareSpout(int messageCount) throws Exception {
+        SingleTopicKafkaUnitSetupHelper.populateTopicData(kafkaUnitRule.getKafkaUnit(), SingleTopicKafkaSpoutConfiguration.TOPIC, messageCount);
+        SingleTopicKafkaUnitSetupHelper.initializeSpout(spout, conf, topologyContext, collector);
+    }
+
+    /**
+     * Helper method to in sequence do:
+     * <li>
+     *     <ul>spout.nexTuple()</ul>
+     *     <ul>verify messageId</ul>
+     *     <ul>spout.ack(msgId)</ul>
+     *     <ul>reset(collector) to be able to reuse mock</ul>
+     * </li>
+     *
+     * @param offset offset of message to be verified
+     * @return {@link ArgumentCaptor} of the messageId verified
+     */
+    ArgumentCaptor<Object> nextTuple_verifyEmitted_ack_resetCollectorMock(int offset) {
```

Comment:
This is the proper way to name tests, please see:
https://stackoverflow.com/questions/155436/unit-test-naming-best-practices. 

Even Google coding style mentions it:
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names

Reply:
Very long camel case names make it very hard to read, and properly asses what is being test.

Reply:
The stackoverflow link isn't about Java, and the answer seems to be talking about C#. The names here also don't follow the pattern outlined in the answer. You're right that the Google style guide allows you to use underscores in test method names.

I don't want to bikeshed you to death, and the test method names are fine, but nextTuple_verifyEmitted_ack_resetCollectorMock isn't a test method, and the name for it could be better I think. It's just emitting a tuple and acking it, why do we need to list out the function names it calls?